zrml/court/src/types.rs

```diff
@@ -0,0 +1,375 @@
+impl VoteItem {
+    pub fn into_outcome(self) -> Option<OutcomeReport> {
+        match self {
+            Self::Outcome(report) => Some(report),
+            _ => None,
+        }
+    }
+
+    pub fn is_outcome(&self) -> bool {
+        matches!(self, Self::Outcome(_))
+    }
+
+    pub fn is_binary(&self) -> bool {
+        matches!(self, Self::Binary(_))
+    }
+}
+
+/// The general information about a particular juror.
+#[derive(
+    parity_scale_codec::Decode,
+    parity_scale_codec::Encode,
+    parity_scale_codec::MaxEncodedLen,
+    scale_info::TypeInfo,
+    Clone,
+    Debug,
+    PartialEq,
+    Eq,
+)]
+pub struct JurorInfo<Balance, BlockNumber, Delegations> {
+    /// The juror's amount in the stake weighted pool.
+    /// This amount is used to find a juror with a binary search on the pool.
+    pub stake: Balance,
+    /// The current amount of funds which are locked in courts.
+    pub active_lock: Balance,
+    /// The block number when a juror exit from court was requested.
+    pub prepare_exit_at: Option<BlockNumber>,
+    pub delegations: Option<Delegations>,
+}
+
+/// The raw information behind the secret hash of a juror's vote.
+pub struct RawCommitment<AccountId, Hash> {
+    /// The juror's account id.
+    pub juror: AccountId,
+    /// The vote item which the juror voted for.
+    pub vote_item: VoteItem,
+    /// The salt which was used to hash the vote.
+    pub salt: Hash,
+}
+
+/// The raw information which is hashed to create the secret hash of a juror's vote.
+pub struct CommitmentMatcher<AccountId, Hash> {
+    /// The juror's hashed commitment
+    pub hashed: Hash,
+    /// The raw commitment which is intended to lead to the hashed commitment.
+    pub raw: RawCommitment<AccountId, Hash>,
+}
+
+/// All possible states of a vote.
+#[derive(
+    parity_scale_codec::Decode,
+    parity_scale_codec::Encode,
+    parity_scale_codec::MaxEncodedLen,
+    scale_info::TypeInfo,
+    Clone,
+    Debug,
+    PartialEq,
+    Eq,
+)]
+pub enum Vote<Hash, DelegatedStakes> {
+    /// The juror delegated stake to other jurors.
+    Delegated { delegated_stakes: DelegatedStakes },
+    /// The juror was randomly selected to vote in a specific court case.
+    Drawn,
+    /// The juror casted a vote, only providing a hash, which meaning is unknown.
+    Secret { commitment: Hash },
+    /// The juror revealed her raw vote, letting anyone know what she voted.
+    Revealed { commitment: Hash, vote_item: VoteItem, salt: Hash },
+    /// The juror was denounced, because she revealed her raw vote during the vote phase.
+    Denounced { commitment: Hash, vote_item: VoteItem, salt: Hash },
+}
+
+/// The information about the lifecycle of a court case.
+#[derive(
+    parity_scale_codec::Decode,
+    parity_scale_codec::Encode,
+    parity_scale_codec::MaxEncodedLen,
+    scale_info::TypeInfo,
+    Clone,
+    Debug,
+    PartialEq,
+    Eq,
+)]
+pub struct CycleEnds<BlockNumber> {
```

Comment:
I understand your confusion here. It was based on one of your previous review comments. Because you mentioned there is a difference between a duration and an end block. So I used a new struct for durations called `RoundTiming`. The `CycleEnds` are just determining the end blocks of those durations.

> pre_vote [...] I would have expected the field in RoundTiming to be a duration, based on the other fields.

Yeah that's a common expectation, I agree.

> Also confused by the fact that all fields have the same name, except for pre_vote vs. pre_vote_end.

That is also true. For `CycleEnds` I had `Ends` already in the name of the struct, so I didn't bother to write it in every field. For `RoundTiming` I wanted to make it clear that it's not a duration, but an end time, which is required, because `pre_vote` is never a period (it is a block number in the future when voting happens `RequestBlock`).

> What's the difference between cycle and round?

There is no difference. Just another name for the same meaning.

Reply:
1fd7d28c1d831187c52786ce3a3cf0f5c86313e6

Do you think this is sufficient?